File: cpp/2021.brightest-position-on-street.cpp
class Solution {
public:
    int solve(vector<vector<int>>& lights) {
	    int n = lights.size();
	    map<long long, int> freq;
	    for(auto &light:lights){
		    freq[light[0]-light[1]]++;
		    freq[light[0]+light[1]+1]--;
	    }
	    long long current = 0, res = 0, start = INT_MAX;
	    for(auto &[key, val]:freq){
		    current +=val;
		    if(current > res){
			    current = res;
			    start = key;
		    }else if(current == res){
			    start = min(start, key);
		    }
	    }
	    return start;
    }
    int brightestPosition(vector<vector<int>>& lights) {
	    return solve(lights);
    }
};

File: python/2021.brightest-position-on-street.py
class Solution:
    def brightestPosition(self, lights: List[List[int]]) -> int:
        intervals = sortedcontainers.SortedDict()
        for light in lights:
            l, r = light[0]-light[1], light[0]+light[1]+1
            intervals[l] = intervals.get(l,0)+1
            intervals[r] = intervals.get(r,0)-1
        current = 0
        res = -math.inf
        start = -math.inf
        for key,val in intervals.items():
            current += val
            if res < current:
                res = current
                start = key
            elif res == current:
                start = min(start, key)
        return start
